src/lib/net/FingerprintDatabase.cpp

```diff
@@ -0,0 +1,128 @@
+/*
+ * Deskflow -- mouse and keyboard sharing utility
+ * SPDX-FileCopyrightText: (C) 2025 Deskflow Developers
+ * SPDX-FileCopyrightText: (C) 2021 Barrier Contributors
+ * SPDX-License-Identifier: GPL-2.0-only WITH LicenseRef-OpenSSL-Exception
+ */
+
+#include "FingerprintDatabase.h"
+
+#include "base/String.h"
+#include "io/filesystem.h"
+
+#include <algorithm>
+#include <fstream>
+
+namespace deskflow {
+
+void FingerprintDatabase::read(const fs::path &path)
+{
+  std::ifstream file;
+  openUtf8Path(file, path, std::ios_base::in);
+  readStream(file);
+}
+
+void FingerprintDatabase::write(const fs::path &path)
+{
+  std::ofstream file;
+  openUtf8Path(file, path, std::ios_base::out);
+  writeStream(file);
+}
+
+void FingerprintDatabase::readStream(std::istream &stream)
+{
+  if (!stream.good()) {
+    return;
+  }
+
+  std::string line;
+  while (std::getline(stream, line)) {
+    if (line.empty()) {
+      continue;
+    }
+
+    auto fingerprint = parseDbLine(line);
+    if (!fingerprint.valid()) {
+      continue;
+    }
+
+    m_fingerprints.push_back(fingerprint);
+  }
+}
+
+void FingerprintDatabase::writeStream(std::ostream &stream)
+{
+  if (!stream.good()) {
+    return;
+  }
+
+  for (const auto &fingerprint : m_fingerprints) {
+    stream << toDbLine(fingerprint) << "\n";
+  }
+}
+
+void FingerprintDatabase::clear()
+{
+  m_fingerprints.clear();
+}
+
+void FingerprintDatabase::addTrusted(const FingerprintData &fingerprint)
+{
+  if (isTrusted(fingerprint)) {
+    return;
+  }
+  m_fingerprints.push_back(fingerprint);
+}
+
+bool FingerprintDatabase::isTrusted(const FingerprintData &fingerprint)
+{
+  auto found = std::find(m_fingerprints.begin(), m_fingerprints.end(), fingerprint);
+  return found != m_fingerprints.end();
+}
+
+FingerprintData FingerprintDatabase::parseDbLine(const std::string &line)
+{
+  FingerprintData result;
+
+  // legacy v1 certificate handling
+  if (std::count(line.begin(), line.end(), ':') == 19 && line.size() == 40 + 19) {
```

Comment:
I know this is ported code but magic numbers are frustrating. Every future dev is going to waste time asking what do `19` and `40` mean? I think it'd be helpful to explain them.

```suggestion
  const auto kSha1ColonCount = 19;
  const auto kSha1HexCharCount = 40;
  const auto kSha1ExpectedSize = kSha1HexCharCount + kSha1ColonCount;
  if (std::count(line.begin(), line.end(), ':') == kSha1ColonCount && line.size() == kSha1ExpectedSize) {
```